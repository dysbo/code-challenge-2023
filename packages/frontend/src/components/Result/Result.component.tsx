import { Link, TableCell, TableRow } from '@mui/material'
import { useCallback, useMemo } from 'react'
import { useAppContext } from '../../hooks'
import { ResultProps } from './Result.type'

export const Result = ({ input, inputTimesTwo, inputTimesTwoSquared, removeResult }: ResultProps) => {
  const { result } = useAppContext()
  const handleRemove = useCallback(() => removeResult(input), [input, removeResult])
  const isActive = useMemo(() => input === result?.input, [result])

  return (
    <TableRow
      sx={{
        backgroundColor: isActive ? 'primary.light' : 'inherit'
      }}
    >
      <TableCell data-testid={`result-n-${input}`}>{input}</TableCell>
      <TableCell data-testid={`result-nTimesTwo-${input}`}>{inputTimesTwo}</TableCell>
      <TableCell data-testid={`result-nTimesTwoSquared-${input}`}>{inputTimesTwoSquared}</TableCell>
      <TableCell data-testid={`result-actions-${input}`}>
        <Link href="#" color="error" onClick={handleRemove}>
          Remove
        </Link>
      </TableCell>
    </TableRow>
  )
}
