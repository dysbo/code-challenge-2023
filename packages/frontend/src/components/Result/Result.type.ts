import { AppContextResult } from '../../providers/AppContextProvider/AppContext.type'

export interface ResultProps extends AppContextResult {
  removeResult: (input: string) => void
}
