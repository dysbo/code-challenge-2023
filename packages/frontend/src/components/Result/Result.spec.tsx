import { act, fireEvent, render } from '@testing-library/react'
import { PropsWithChildren } from 'react'
import { Table, TableBody } from '@mui/material'
import { appContext } from '../../providers'
import { Result } from './Result.component'

// eslint-disable-next-line react/display-name
const wrapper = (activeNumber?: number) => ({ children }: PropsWithChildren) => (
  <appContext.Provider
    value={{
      clearResult: jest.fn(),
      input: activeNumber?.toString() ?? '',
      onSubmit: jest.fn(),
      result: { input: activeNumber?.toString() ?? '', inputTimesTwo: '', inputTimesTwoSquared: '' },
      setError: jest.fn(),
      setInput: jest.fn()
    }}>
    <Table>
      <TableBody>
        {children}
      </TableBody>
    </Table>
  </appContext.Provider>
)

describe('Result tests', () => {
  test('should show result when it is provided', () => {
    const { container } = render(<Result removeResult={jest.fn()} input="3" inputTimesTwo="6"
      inputTimesTwoSquared="36" />, { wrapper: wrapper(3) })
    expect(container).toMatchSnapshot()
  })

  test('should call to remove row when remove is clicked', async () => {
    const mockRemoveResult = jest.fn()
    const { findByTestId } = render(<Result removeResult={mockRemoveResult} input='4' inputTimesTwo='16' inputTimesTwoSquared='256' />, { wrapper: wrapper() })

    const btnRemove = await findByTestId('result-actions-4-remove')

    act(() => {
      fireEvent.click(btnRemove)
    })

    expect(mockRemoveResult).toHaveBeenCalledTimes(1)
  })
})