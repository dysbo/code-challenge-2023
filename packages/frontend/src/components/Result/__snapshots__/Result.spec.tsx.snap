// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Result tests should show result when it is provided 1`] = `
<div>
  <table
    class="MuiTable-root css-rqglhn-MuiTable-root"
  >
    <tbody
      class="MuiTableBody-root css-apqrd9-MuiTableBody-root"
    >
      <tr
        class="MuiTableRow-root css-bwjzc7-MuiTableRow-root"
      >
        <td
          class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
          data-testid="result-n-3"
        >
          3
        </td>
        <td
          class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
          data-testid="result-nTimesTwo-3"
        >
          6
        </td>
        <td
          class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
          data-testid="result-nTimesTwoSquared-3"
        >
          36
        </td>
        <td
          class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
          data-testid="result-actions-3"
        >
          <a
            class="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineAlways css-l69lu0-MuiTypography-root-MuiLink-root"
            data-testid="result-actions-3-remove"
            href="#"
          >
            Remove
          </a>
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
